lib/services/local/dynamic_access.go

```diff
@@ -158,6 +176,138 @@ func (s *DynamicAccessService) UpsertAccessRequest(ctx context.Context, req serv
 	return nil
 }
 
+// GetPluginData loads all plugin data matching the supplied filter.
+func (s *DynamicAccessService) GetPluginData(ctx context.Context, filter services.PluginDataFilter) ([]services.PluginData, error) {
+	switch filter.Kind {
+	case services.KindAccessRequest:
+		data, err := s.getAccessRequestPluginData(ctx, filter)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return data, nil
+	default:
+		return nil, trace.BadParameter("unsupported resource kind %q", filter.Kind)
+	}
+}
+
+func (s *DynamicAccessService) getAccessRequestPluginData(ctx context.Context, filter services.PluginDataFilter) ([]services.PluginData, error) {
+	// Filters which specify Resource are a special case since they will match exactly zero or one
+	// possible PluginData instances.
+	if filter.Resource != "" {
+		item, err := s.Get(ctx, pluginDataKey(services.KindAccessRequest, filter.Resource))
+		if err != nil {
+			if trace.IsNotFound(err) {
+				// Caller is going to need to check the length of the returned slice anyway,
+				// so propagating a NotFoundError is redundant.
```

Comment:
Disagree. So far in teleport and all other products, we never have 3 state problem:

* State 1: Object is correct, nil
* State 2: Object is incorrect, err
* State 3: nil, nil <- this third state is redundant, because every single caller now has to verify for thirds logical branch. 
If you only have State 1 and State 2, the code safely assumes, that if `err == nil` the object is valid. This part violates this invariant.

Reply:
I initially agreed with this comment, and implemented appropriate changes, but that caused other errors which made me look closer... Turns out that we do in fact return `nil, nil` all the time!  We just don't generally do it explicitly.  Take, for example, this simplified version of the source of `GetUsers`:

```golang
result, _ := s.GetRange(...)
var out []services.User
for _, item := range result.Items {
	// do some conversions
	out = append(out, user)
}
return out, nil
```

If no users exist, `nil, nil` is in fact the return value of the function, and the rest of `teleport` handles this correctly.

The only reason why `nil, nil` is being returned *explicitly* in this PR is due to a performance optimization which short-circuits queries which match exactly zero or one resources.

Reply:
true, everything I said was meant for scalar values, for lists, returning a list of len 0 or nil is OK, is that this the case?

Reply:
Yep, we're returning a slice here.